lib/link.ex

```diff
@@ -42,35 +49,182 @@ defmodule Link do
+        project = Enum.at(parts, 1)
+        pr_number = Enum.at(parts, 3)
+
+        # Assemble into the compact form:
+        "#{org}/#{project}/PR##{pr_number}"
+
+      true ->
+        # Orgs, People or Repos just return path e.g: "dwyl/app" or "iteles"
+        clean
+    end
+  end
+
+  @doc """
+  `strip_protocol/1` strips the protocol e.g: "https://" from a URL.
+
+  ## Examples
 
-      # Assemble into the compact form:
-      "#{org}/#{project}##{issue_number}"
+      iex> Link.strip_protocol("https://dwyl.com")
+      "dwyl.com"
+  """
+  def strip_protocol(url) do
+    String.replace(url, ~r"(http(s)?):\/\/", "")
+  end
+
+  @doc """
+  `strip_trailing_slash/1` strips trailing forward slash from URL.
 
+  ## Examples
+
+      iex> Link.strip_trailing_slash("dwyl.com/")
+      "dwyl.com"
+  """
+  def strip_trailing_slash(url) do
+    if String.ends_with?(url, "/") && url != "/" do
+      String.slice(url, 0..-2)
     else
-      # Orgs, People or Repos just return the path e.g: "dwyl/app" or "iteles"
-      clean
+      url
     end
   end
 
-  def remove_protocol(url) do
-    url
-    |> String.replace("https://", "")
-    |> String.replace("http://", "")
+  @doc """
+  `regex/0` returns the Regular Expression needed to match URLs.
+  According to `RFC 3986`: https://www.rfc-editor.org/rfc/rfc3986
+  Based on reading https://mathiasbynens.be/demo/url-regex
+  After much searching on Google, GitHub and StackOverflow,
+  this is what we came up with.
+
+  #HELPWANTED: if you find a *better* (faster, more comppliant) RegEx
+  that passes all our tests, please share! github.com/dwyl/link/issues/new
+
+  Explanation:
+
+  ((http(s)?):\/\/             # Optional scheme (http or https)
+  (www\.)?                     # Optional "www"
+  a-zA-Z0-9@:%._\+~#=]{2,256}  # Domain (IPv4, IPv6 or hostname)
+  :%                           # Optional port number
+  [a-z]{2,6}\b                 # Domain extension e.g: ".com"
+  (?:\?[^\n\r]*)?              # Optional query string ?q=
+  (?:#[^\n\r]*)?               # Optional fragment e.g: #comment
+
+  ## Examples
+
+      iex> Regex.run(Link.regex(), "dwyl.com") |> List.flatten |> Enum.filter(& &1 != "") |> List.first
+      "dwyl.com"
+  """
+  def regex do
+    ~r|[(http(s)?):\/\/(www\.)?a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,14}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)|
+  end
+
+  @doc """
+  `valid?/1` confirms if a URL is valid
+  using the `RFC 3986` compliant `regex/0` above.
+
+  ## Examples
+
+      iex> Link.valid?("example.c")
+      false
+
+      iex> Link.valid?("https://www.example.com")
+      true
+  """
+  def valid?(url) do
+    Regex.match?(regex(), url)
+  end
+
+  @doc """
+  `find/1` finds all instances of a URL in a block of text.
+
+  ## Examples
+
+      iex> Link.find("Text with links http://goo.gl/3co4ae and https://git.io/top & www.dwyl.com etc.")
+      ["http://goo.gl/3co4ae", "https://git.io/top", "www.dwyl.com"]
+
+      # Find a link without any other text or whitespace #8
```

Comment:
Added this doctest for #8 
Passes without modifying any cost

Reply:
PR links are working. https://github.com/dwyl/link/pull/5#issuecomment-1666943043 ✅